tree/dataframe/inc/ROOT/RDF/RVariedAction.hxx

```diff
@@ -76,21 +76,43 @@ class R__CLING_PTRCHECK(off) RVariedAction final : public RActionBase {
    }
 
 public:
-   RVariedAction(std::vector<Helper> &&helpers, const ColumnNames_t &columns, std::shared_ptr<PrevNode> prevNode,
-                 const RColumnRegister &colRegister)
-      : RActionBase(prevNode->GetLoopManagerUnchecked(), columns, colRegister, prevNode->GetVariations()),
-        fHelpers(std::move(helpers)), fPrevNodes(MakePrevFilters(prevNode)), fInputValues(GetNSlots())
+   void SetupClass()
    {
+      // The column register and names are private members of RActionBase
+      const auto &colRegister = GetColRegister();
+      const auto &columnNames = GetColumnNames();
+
       fLoopManager->Register(this);
 
-      for (auto i = 0u; i < columns.size(); ++i) {
-         auto *define = colRegister.GetDefine(columns[i]);
+      for (auto i = 0u; i < columnNames.size(); ++i) {
+         auto *define = colRegister.GetDefine(columnNames[i]);
          fIsDefine[i] = define != nullptr;
          if (fIsDefine[i])
             define->MakeVariations(GetVariations());
       }
    }
 
+   RVariedAction(std::vector<Helper> &&helpers, const ColumnNames_t &columns, std::shared_ptr<PrevNode> prevNode,
+                 const RColumnRegister &colRegister)
+      : RActionBase(prevNode->GetLoopManagerUnchecked(), columns, colRegister, prevNode->GetVariations()),
+        fHelpers(std::move(helpers)),
+        fPrevNodes(MakePrevFilters(prevNode)),
+        fInputValues(GetNSlots())
+   {
+      SetupClass();
+   }
+
+   /// This constructor takes in input a vector of previous filter nodes, motivated by the CloneAction logic.
+   RVariedAction(std::vector<Helper> &&helpers, const ColumnNames_t &columns,
+                 const std::vector<std::shared_ptr<PrevNodeType>> &prevFilters, const RColumnRegister &colRegister)
```

Comment:
```suggestion
                 const std::vector<std::shared_ptr<PrevNodeType>> &prevNodes, const RColumnRegister &colRegister)
```

they might not be filters (could be ranges or copies of the rloopmanager)

Reply:
I see. I was a bit confused due to the usage of the `prevNode` argument of the main class constructor:
https://github.com/root-project/root/blob/cbbfeb74da355229fd993db7535fbf658e0c3179/tree/dataframe/inc/ROOT/RDF/RVariedAction.hxx#L87

Since it's used directly as the input argument to `MakePrevFilters` I thought it was a filter itself. But actually the function then either creates a vector with the loop manager or calls `prevNode->GetVariedFilter` which is implemented both for `RFilter` and `RRange`. All clear now, thanks

Reply:
I see, I can't come up with a better name for `MakePrevFilters` on the spot (because _mostly_ that's the interesting thing it does) but feel free to propose one!

Reply:
uhm, maybe `GetVariedPrevNodes`?

Reply:
I would go with `MakeVariedPrevNodes`

Reply:
:+1: 

for another PR I guess